packages/ui-components/src/components/toolbar.tsx

```diff
@@ -223,12 +223,26 @@ export class Toolbar<T extends Widget = Widget> extends Widget {
     if (existing) {
       return false;
     }
-    widget.addClass(TOOLBAR_ITEM_CLASS);
     const layout = this.layout as ToolbarLayout;
 
     const j = Math.max(0, Math.min(index, layout.widgets.length));
     layout.insertWidget(j, widget);
 
+    // Wait for the ReactWidget to be rendered before adding class its class.
+    if (
+      widget instanceof ReactWidget &&
+      (widget as ReactWidget).renderPromise !== undefined
+    ) {
+      (widget as ReactWidget)
+        .renderPromise!.then(() => {
+          widget.addClass(TOOLBAR_ITEM_CLASS);
+        })
+        .catch(() => {
+          console.error(`Error loading toolbar item ${name}`);
+        });
+    } else {
+      widget.addClass(TOOLBAR_ITEM_CLASS);
+    }
```

Comment:
I'm surprised by this as the node on which the class is added exists before the React component is mounted in it. Or is React removing the classes on its root?

Reply:
Looks like the test still fails, you were right. It was working locally, probably it takes a bit more time to execute locally and the widget is rendered.

Another solution could be to wait for all the buttons to be rendered before attaching the toolbar to the DOM here:
https://github.com/jupyterlab/jupyterlab/blob/6d208970f0b950a1c75c0c08960c651b32a3cc0b/packages/cell-toolbar/src/celltoolbartracker.ts#L112-L117
Again it works locally, and should be more robust as this is triggered every time the selected cell changes. But this adds some code only for the tests...

Or we can also add a timeout in the tests...

Reply:
There is a helper for the test `await framePromise` to wait for a frame refresh. Sometimes it is enough to make a test more robust.

Reply:
Thanks, I'll try it and revert these changes

Reply:
But it seems that all Jest unit tests are green without it.

Reply:
Yes, it's about UI-test on collapsible heading, one does not work because it expect a toolbar in the snapshot which is not ready when the snapshot is captured.
So the `framePromise` will not be an option finally.

Reply:
The test which fails switches quickly between the cells, making the buttons (`ReactWidget`) to render/unmount quickly.
It seems to break something, probably because of the new API using `createRoot` in the `vdom` package.
I suspect the widget is detached while the root is creating.